### ISSUE TITLE
feat(merge): MoE construction from dense models (Mixtral-style)

### ISSUE THREAD

**Issue by noahgift** [paiml/aprender#445]
## Summary

Combine N dense models into a Mixture-of-Experts architecture. Each dense
model becomes an expert. A router layer is added and trained on calibration
data to select experts per token.

## CLI

```bash
apr merge moe \
  --experts model-a.apr model-b.apr model-c.apr model-d.apr \
  --router-data calibration.jsonl \
  --num-experts-per-token 2 \
  --output-format mixtral \
  -o moe-model.apr
```

## Parameters

- `--experts`: 2+ dense models to combine as experts
- `--router-data`: Calibration data for router training
- `--num-experts-per-token`: Active experts per token (default: 2)
- `--output-format`: MoE architecture format (mixtral, deepseek, qwen)
- `--router-type`: Router architecture (top-k, expert-choice)

## Implementation

1. Validate all expert models share compatible architecture (same hidden_size)
2. Extract MLP layers from each expert model
3. Construct MoE layer structure with router
4. Train router on calibration data (small gradient descent)
5. Assemble final model with shared attention + expert MLPs

## MergeKit Reference

MergeKit supports Mixtral, DeepSeek MoE, and Qwen MoE output formats.

## Specification

See `docs/specifications/components/merge.md` §4.9.

### Contract: `moe-construction-v1.yaml`

```yaml
metadata:
  description: "MoE construction from dense models — expert assembly + router training"
  references:
    - "Shazeer et al. (2017) Outrageously Large Neural Networks (MoE)"
    - "Jiang et al. (2024) Mixtral of Experts"
  depends_on:
    - "softmax-kernel-v1"

equations:
  router_output:
    formula: "g(x) = TopK(softmax(W_r · x), k)"
    invariants:
      - "Exactly k experts selected per token"
      - "Router weights sum to 1 after TopK renormalization"
      - "Output shape = hidden_size (same as each expert)"

  expert_assembly:
    formula: "MoE_layer = {shared_attn, expert_mlp_1, ..., expert_mlp_n, router}"
    invariants:
      - "All experts have identical MLP architecture"
      - "Shared attention layer from base model"
      - "n_experts = number of input models"

  load_balance:
    formula: "balance_loss = α * n * Σ f_i · P_i"
    invariants:
      - "f_i = fraction of tokens routed to expert i"
      - "P_i = mean router probability for expert i"
      - "Minimized during router training"

proof_obligations:
  - type: invariant
    property: "Expert count matches input models"
    formal: "moe.n_experts == len(input_models)"
  - type: invariant
    property: "TopK selection count"
    formal: "count(selected_experts) == k for every token"
  - type: invariant
    property: "Shape preservation through MoE layer"
    formal: "output.shape == input.shape"
  - type: conservation
    property: "Router weight normalization"
    formal: "sum(router_weights[selected]) == 1.0"

falsification_tests:
  - id: FALSIFY-MOE-001
    rule: "Expert count"
    prediction: "MoE model has exactly N experts for N input models"
    if_fails: "Expert extraction or assembly bug"
  - id: FALSIFY-MOE-002
    rule: "TopK invariant"
    prediction: "Exactly k experts active per forward pass"
    if_fails: "Router TopK implementation broken"
  - id: FALSIFY-MOE-003
    rule: "Shape preservation"
    prediction: "MoE layer output shape == input shape"
    if_fails: "Expert output concatenation or gating bug"
  - id: FALSIFY-MOE-004
    rule: "Architecture compatibility"
    prediction: "incompatible hidden_size models produce validation error"
    if_fails: "Missing compatibility check"
```

### Binding Requirements

Add to `contracts/aprender/binding.yaml`:
```yaml
  - contract: moe-construction-v1.yaml
    equation: router_output
    module_path: "aprender::format::moe"
    function: construct_moe
    status: not_implemented
```

### Verification Targets
- L2: 4 falsification tests
- L3: Kani harness for TopK selection count (bound k≤4, n≤8)
- L4: Lean proof for router weight normalization (reuse softmax partition_of_unity)
